### PR TITLE
feat: add WooCrossChainRouterV2

### PR DESCRIPTION
**Pull request by merlincdj**

### CODE REVIEW

File: contracts/interfaces/IWooCrossChainRouterV2.sol [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+import "../interfaces/IWooPPV2.sol";
```
Comment (merlincdj) [inline contracts/interfaces/IWooCrossChainRouterV2.sol:37]: will remove this useless import next commit.

File: contracts/WooCrossChainRouterV2.sol [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+// Local Contracts
+import "./interfaces/IWooCrossChainRouterV2.sol";
+import "./interfaces/IWooRouterV2.sol";
+import "./interfaces/IWETH.sol";
+import "./interfaces/Stargate/IStargateRouter.sol";
+import "./interfaces/Stargate/IStargateReceiver.sol";
```
Comment (merlincdj) [inline contracts/WooCrossChainRouterV2.sol:16]: will remove this useless import next commit.

Comment (merlincdj) [inline contracts/WooCrossChainRouterV2.sol:16]: will remove this useless import next commit.

File: contracts/interfaces/IWooCrossChainRouterV2.sol (hunk 1 above)
Comment (merlincdj) [inline contracts/interfaces/IWooCrossChainRouterV2.sol:37]: will remove this useless import next commit.

File: contracts/WooCrossChainRouterV2.sol [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+    address public constant ETH_PLACEHOLDER_ADDR = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /* ----- Variables ----- */
+
+    IWooRouterV2 public wooRouter;
+    IStargateRouter public stargateRouter;
+
+    address public override weth;
+    // address public quoteToken;
```
Comment (merlincdj) [inline contracts/WooCrossChainRouterV2.sol:36]: every comment like this like will remove next commit.

File: contracts/WooCrossChainRouterV2.sol [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+                IWETH(srcInfos.fromToken).deposit{value: srcInfos.fromAmount}();
+                msgValue -= srcInfos.fromAmount;
+            } else {
+                TransferHelper.safeTransferFrom(srcInfos.fromToken, msg.sender, address(this), srcInfos.fromAmount);
+            }
+
+            // Step 2: local swap by WooRouter or not
+            // 1.WOO is directBridgeToken, path(always) WOO(Arbitrum) => WOO(BSC)
+            // 2.WOO not the directBridgeToken, path(maybe): WOO(Arbitrum) -> ETH(Arbitrum) => ETH(BSC) -> WOO(BSC)
+            // 3.Ethereum no WOOFi liquidity, tokens(WOO, ETH, USDC) always will be direct bridged without swap
```
Comment (merlincdj) [inline contracts/WooCrossChainRouterV2.sol:116]: typo, `direct bridged` -> `bridged directly`

File: contracts/WooCrossChainRouterV2.sol [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+                    toToken,
+                    ETH_PLACEHOLDER_ADDR,
+                    minToAmount,
+                    amountLD
+                );
+            }
+        } else {
+            // Bridged by ERC20 token, holding ERC20 token assets, require swap to toToken(can be ETH)
+            // TODO: Discuss this situation: bridgedToken(BUSD), toToken(USDT), not support stable coin swap now on WooRouter
+            // Waste gas by code below?
```
Comment (merlincdj) [inline contracts/WooCrossChainRouterV2.sol:252]: need discuss here.

File: contracts/WooCrossChainRouter.sol [hunk 6]
```diff
@@ -1,372 +1,446 @@
...
+
+        address msgSender = _msgSender();
+        uint256 msgValue = msg.value;
+        uint256 bridgeAmount;
 
-    function setDstGasForNoSwapCall(uint256 _dstGasForNoSwapCall) external onlyOwner {
-        dstGasForNoSwapCall = _dstGasForNoSwapCall;
-    }
+        {
+            if (srcInfos.bridgeToken == ETH_PLACEHOLDER_ADDR) {
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:92]: Is this allowed to be set as "ETH_PLACEHOLDER_ADDR" ?

File: contracts/WooCrossChainRouter.sol [hunk 7]
```diff
@@ -1,372 +1,446 @@
...
-                _token,
+                ETH_PLACEHOLDER_ADDR,
                 minToAmount,
                 amountLD
             );
             return;
         }
 
-        uint256 quoteAmount = amountLD;
+        if (toToken == bridgedToken) {
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:205]: is that possible here, the bridgedToken could be ETH_PLACEHOLDER_ADDR ?

File: contracts/WooCrossChainRouter.sol [hunk 8]
```diff
@@ -1,372 +1,446 @@
...
+        sgPoolIds[112][0x04068DA6C83AFCFA0e13ba15A6696662335D5B75] = 1; // USDC
+    }
+
+    function _getLzTxObj(address to, DstInfos memory dstInfos) internal view returns (IStargateRouter.lzTxObj memory) {
+        uint256 dstGasForCall = (dstInfos.toToken == dstInfos.bridgeToken) ? dstGasForNoSwapCall : dstGasForSwapCall;
+
+        return IStargateRouter.lzTxObj(dstGasForCall, dstInfos.airdropNativeAmount, abi.encodePacked(to));
+    }
+
+    function _encodePayload(
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:361]: if this function is only one line, probably won't need to be defined as an internal function, right?

File: contracts/WooCrossChainRouter.sol [hunk 9]
```diff
@@ -1,372 +1,446 @@
...
+        return
+            abi.encode(
+                refId, // reference id
+                to, // address for receive tokens
+                dstInfos.toToken, // to token on dst chain
+                dstInfos.minToAmount // minToAmount on dst chain
+            );
+    }
+
+    function _decodePayload(bytes memory payload)
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:375]: ditto

**Review by fb-alexcq — CHANGES_REQUESTED** [review]
Several inline comments.

File: contracts/WooCrossChainRouter.sol (hunk 8 above)
Comment (merlincdj) [reply contracts/WooCrossChainRouter.sol:361]: Right, now two functions `crossSwap` & `quoteLayerZeroFee` have same encode logic, defined as internal function just for reduce the code duplicate.

File: contracts/WooCrossChainRouter.sol (hunk 7 above)
Comment (merlincdj) [reply contracts/WooCrossChainRouter.sol:205]: In `sgReceive`, bridgedToken wont be `ETH_PLACEHOLDER_ADDR`. Only ERC20 token address.

File: contracts/WooCrossChainRouter.sol (hunk 6 above)
Comment (merlincdj) [reply contracts/WooCrossChainRouter.sol:92]: `bridgeToken` could be `USDC/USDT/BUSD`, not only `ETH_PLACEHOLDER_ADDR`. Have a look on this function `_initSGPoolIds`.

File: contracts/WooCrossChainRouter.sol (hunk 7 above)
Comment (merlincdj) [reply contracts/WooCrossChainRouter.sol:205]: If bridged by ETH on source chain, then sgRecevie`s bridgedToken will be [SGETH](https://arbiscan.io/address/0x82CbeCF39bEe528B5476FE6d1550af59a9dB6Fc0) address.

File: contracts/WooCrossChainRouter.sol (hunk 9 above)
Comment (merlincdj) [reply contracts/WooCrossChainRouter.sol:375]: agree

File: contracts/WooCrossChainRouter.sol [hunk 10]
```diff
@@ -1,372 +1,435 @@
...
+                to,
+                bridgedToken,
+                amountLD,
+                toToken,
+                bridgedToken,
+                minToAmount,
+                amountLD
+            );
+            return;
+        }
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:224]: if (bridgedToken == sgETHs[localId]) {
   // sgReceived native
   if (to == PLACEHOLDER) {
      safeTransferETH(...);
   } else {
      wooRouter.swap{value: amountLD};
   }
} else {
   // sgReceived ERC20
   if (....) {
      safeTransfer()
   } else {
     wooRouter.swap
   }
}

File: contracts/WooCrossChainRouter.sol [hunk 11]
```diff
@@ -278,10 +252,40 @@ contract WooCrossChainRouterV2 is IWooCrossChainRouterV2, Ownable, ReentrancyGua
...
                     toToken,
-                    bridgedToken,
+                    toToken,
                     minToAmount,
                     amountLD
                 );
+            } else {
+                TransferHelper.safeApprove(bridgedToken, address(wooRouter), amountLD);
+                try wooRouter.swap(bridgedToken, toToken, amountLD, minToAmount, payable(to), to) returns (
+                    uint256 realToAmount
```
Comment (fb-alexcq) [inline contracts/WooCrossChainRouter.sol:262]: sorry to bother, I didn't remember very clearly . Want to double check:

   realToAmount can be captured as the return value?

**Review by fb-alexcq — APPROVED** [review]
LGTM

**Pull request merged**